File: problems/547-Number-of-Provinces/solution.go
func traverse(city int, isVisited *[]bool, isConnected *[][]int) {
	(*isVisited)[city] = true

	for i := 0; i < len(*isConnected); i++ {
		if (*isConnected)[city][i] == 1 && !(*isVisited)[i] {
			traverse(i, isVisited, isConnected)
		}
	}
}

func findCircleNum(isConnected [][]int) int {
	cityNum, provinceNum := len(isConnected), 0
	isVisited := make([]bool, cityNum)
	for i := 0; i < cityNum; i++ {
		if !isVisited[i] {
			traverse(i, &isVisited, &isConnected)
			provinceNum++
		}
	}

	return provinceNum
}

File: problems/547-Number-of-Provinces/solution.java
class Solution {
    public void traverse(int idx, int[][] isConnected, boolean[] isVisited) {
        isVisited[idx] = true;

        for (int i = 0; i < isConnected[idx].length; i++) {
            if (isConnected[idx][i] == 1 && !isVisited[i]) {
                traverse(i, isConnected, isVisited);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        boolean[] isVisited = new boolean[n];

        int result = 0;
        for (int i = 0; i < n; i++) {
            if (!isVisited[i]) {
                traverse(i, isConnected, isVisited);
                result++;
            }
        }

        return result;
    }
}

